static/js/tab_bar.js

```diff
@@ -1,172 +1,137 @@
-
-            tabs.push(make_tab("Group Private", '#narrow/group-pm-with',
-                               undefined, 'private_message '));
-
-
-        } else if (filter.has_operand("is", "starred")) {
-            tabs.push(make_tab("Starred", hashed));
-        } else if (filter.has_operand("streams", "public")) {
-            tabs.push(make_tab("Public Streams", hashed));
-        } else if (filter.has_operator("near")) {
-            tabs.push(make_tab("Near " + filter.operands("near")[0], hashed));
-        } else if (filter.has_operator("id")) {
-            tabs.push(make_tab("ID " + filter.operands("id")[0], hashed));
-        } else if (filter.has_operand("is", "mentioned")) {
-            tabs.push(make_tab("Mentions", hashed));
-        } else if (filter.has_operator("sender")) {
-            let sender = filter.operands("sender")[0];
-            if (people.get_by_email(sender)) {
-                sender = people.get_by_email(sender).full_name;
-            }
-            tabs.push(make_tab("Sent by " + sender, hashed));
-        }  else if (filter.has_operator("search")) {
-            // Search is not a clickable link, since we don't have
-            // a search narrow
-            tabs.push(make_tab("Search results", false));
-        }
-
-        // Third breadcrumb item for stream-topic naarrows
-        if (filter.has_operator("stream") &&
-            filter.has_operator("topic")) {
-            const topic = filter.operands("topic")[0];
-            hashed = hash_util.operators_to_hash(ops.slice(0, 2));
+exports.colorize_tab_bar = function () {
+    const filter = narrow_state.filter();
+    if (filter === undefined || !filter.has_operator('stream')) {return;}
+    const color_for_stream = stream_data.get_color(filter.operands("stream")[0]);
+    const stream_light = colorspace.getHexColor(colorspace.getDecimalColor(color_for_stream));
+    $("#tab_list .fa-hashtag").css('color', stream_light);
+    $("#tab_list .fa-lock").css('color', stream_light);
+};
 
-            tabs.push(make_tab(topic, hashed, null));
-        }
+function display_tab_bar(tab_bar_data) {
+    const tab_bar = $("#tab_bar");
+    tab_bar.empty();
+    const rendered = render_tab_bar(tab_bar_data);
+    tab_bar.append(rendered);
+    if (tab_bar_data.stream_settings_link) {
+        exports.colorize_tab_bar();
     }
+    tab_bar.removeClass('notdisplayed');
+}
 
-    if (tabs.length === 0) {
-        tabs.push(make_tab('All messages', "#", "home", "root", true));
+function build_tab_bar(filter) {
+    // This makes sure we don't waste time appending tab_bar on a template where it's never used
+    if (filter && !filter.is_common_narrow()) {
+        exports.open_search_bar_and_close_narrow_description();
+    } else {
+        const tab_bar_data = make_tab_data(filter);
+        display_tab_bar(tab_bar_data);
+        click_handlers.bind_handler_for_opening_searchbox();
+        exports.close_search_bar_and_open_narrow_description();
```

Comment:
I think with early-return you could avoid duplicating `exports.close_search_bar_and_open_narrow_description();`

Reply:
I don't think I clearly understand this.